kube/src/api/metadata.rs

```diff
@@ -14,7 +15,28 @@ use serde::{Deserialize, Serialize};
 /// - `.metadata.resource_version`
 ///
 /// This avoids a bunch of the unnecessary unwrap mechanics for apps.
-pub trait Meta: Metadata {
+pub trait Meta {
+    /// Types that know their metadata at compile time should select `Family = ()`.
+    /// Types that require some information at runtime should select `Family`
+    /// as type of this information.
+    type Family: Send + Sync + 'static;
```

Comment:
This is the crux of the PR. We are changing the `Meta` trait to basically be our own, more practical, `Resource` trait. I think that _overall_ is positive; `DynamicResource`, as you pointed out, just could not work in more than trivial situations without this change, and a blanket implementation of our trait of `k8s_openapi::Resource + Metadata` objects is - in theory - easy to do.

A rename of the trait is probably in order to reflect what the trait now does. Not sure to what though.

The `Family` type parameter here is definitely the most awkward implementation detail of the trait though. Ideally, I would like to avoid this, as it's making breaking changes to both `kube::Resource` and `kube::Api` (as well as making them more awkward).

Reply:
Regarding naming, Do you think the following makes sense?
- rename Meta to Resource
- rename Resource to RequestBuilder/QueryBuilder
- rename Family to RuntimeTypeInformation (long name, but anyway users will rarely interact with it)

I tried to make sure that `kube::Api` public API was not modified (if `K` has all necessary information at compile-time, `Api::all` and `Api::namepsaced` are still provided). Did I miss something? I will also add similar backward-compatible constructors to the `Resource`.

Reply:
I think this has a lot more potential now that it's shown that we don't need to break `Api` or `Resource`, although I will leave some comments on the implementation inlined elsewhere.

Let's not worry about the names of those three yet (it'll make it super confusing to review, so we can change them later). Although I think in theory the first two are fine.